dll/win32/shell32/folders/CDrivesFolder.cpp

```diff
@@ -90,27 +93,89 @@ HRESULT CALLBACK DrivesContextMenuCallback(IShellFolder *psf,
+            if (!(dwFlags & FILE_READ_ONLY_VOLUME) && nDriveType != DRIVE_REMOTE)
             {
+                /* add separator and Format */
                 _InsertMenuItemW(pqcminfo->hmenu, pqcminfo->indexMenu++, TRUE, 0, MFT_SEPARATOR, NULL, 0);
                 _InsertMenuItemW(pqcminfo->hmenu, pqcminfo->indexMenu++, TRUE, pqcminfo->idCmdFirst++, MFT_STRING, MAKEINTRESOURCEW(IDS_FORMATDRIVE), MFS_ENABLED);
             }
+            if (nDriveType == DRIVE_CDROM || nDriveType == DRIVE_REMOTE)
+            {
+                /* add separator and Eject */
+                _InsertMenuItemW(pqcminfo->hmenu, pqcminfo->indexMenu++, TRUE, 0, MFT_SEPARATOR, NULL, 0);
+                _InsertMenuItemW(pqcminfo->hmenu, pqcminfo->indexMenu++, TRUE, pqcminfo->idCmdFirst++, MFT_STRING, MAKEINTRESOURCEW(IDS_EJECT), MFS_ENABLED);
+            }
         }
     }
     else if (uMsg == DFM_INVOKECOMMAND)
     {
+        WCHAR wszBuf[4];
+        wcscpy(wszBuf, L"A:\\");
+        wszBuf[0] = (WCHAR)szDrive[0];
+
         if (wParam == DFM_CMD_PROPERTIES)
         {
-            WCHAR wszBuf[4];
-            wcscpy(wszBuf, L"A:\\");
-            wszBuf[0] = (WCHAR)szDrive[0];
             if (!SH_ShowDriveProperties(wszBuf, pidlFolder, apidl))
                 hr = E_FAIL;
         }
         else
         {
-            SHFormatDrive(hwnd, szDrive[0] - 'A', SHFMT_ID_DEFAULT, 0);
+            if (nDriveType == DRIVE_CDROM || nDriveType == DRIVE_REMOTE)
+            {
+                /* Do eject */
+                WCHAR physical[10];
+                wsprintfW(physical, _T("\\\\.\\%c:"), szDrive[0]);
+                HANDLE hDrive = CreateFile(physical, GENERIC_READ | GENERIC_WRITE,
+                                           FILE_SHARE_READ | FILE_SHARE_WRITE, 0, OPEN_EXISTING, 0, NULL);
+                if (hDrive != INVALID_HANDLE_VALUE)
+                {
+                    BOOL bResult;
+                    INT nStrID = 0;
+                    DWORD dwError = NO_ERROR;
+                    do
+                    {
+                        bResult = DeviceIoControl(hDrive, FSCTL_LOCK_VOLUME, NULL, 0, NULL, 0, NULL, NULL);
+                        if (!bResult)
+                        {
+                            nStrID = IDS_CANTLOCKVOLUME; /* Unable to lock volume */
+                            dwError = GetLastError();
+                            break;
+                        }
+                        bResult = DeviceIoControl(hDrive, FSCTL_DISMOUNT_VOLUME, NULL, 0, NULL, 0, NULL, NULL);
+                        if (!bResult)
+                        {
+                            nStrID = IDS_CANTDISMOUNTVOLUME; /* Unable to dismount volume */
+                            dwError = GetLastError();
+                            break;
+                        }
+                        bResult = DeviceIoControl(hDrive, IOCTL_STORAGE_EJECT_MEDIA, NULL, 0, NULL, 0, NULL, NULL);
```

Comment:
You should FSCTL_UNLOCK_VOLUME first.

Reply:
https://support.microsoft.com/en-us/help/165721/how-to-ejecting-removable-media-in-windows-nt-windows-2000-windows-xp
This says it should be locked at first.

Reply:
Ok.